File: packages/gbnf-experiment/data/20260909T005215Z_9a3e20fe/ported_implementation/src/utils/errors/inputParseError.ts
import { buildErrorPosition } from './buildErrorPosition.ts';
import type { ValidInput } from './errorsTypes.ts';
import { getInputAsString } from './getInputAsString.ts';

export const INPUT_PARSER_ERROR_HEADER_MESSAGE = 'Failed to parse input string:';

const buildMessage = (mostRecentInput: ValidInput, pos: number, previousInput: ValidInput): string =>
  [
    INPUT_PARSER_ERROR_HEADER_MESSAGE,
    '',
    ...buildErrorPosition(
      `${getInputAsString(previousInput)}${getInputAsString(mostRecentInput)}`,
      pos + getInputAsString(previousInput).length,
    ),
  ].join('\n');

export class InputParseError extends Error {
  mostRecentInput: ValidInput;
  pos: number;
  previousInput: ValidInput;

  constructor(mostRecentInput: ValidInput, pos: number, previousInput: ValidInput = '') {
    super(buildMessage(mostRecentInput, pos, previousInput));
    this.name = 'InputParseError';
    this.mostRecentInput = mostRecentInput;
    this.pos = pos;
    this.previousInput = previousInput;
  }

  get src(): string {
    return `${getInputAsString(this.previousInput)}${getInputAsString(this.mostRecentInput)}`;
  }

  get errorForMostRecentInput(): string {
    return [
      INPUT_PARSER_ERROR_HEADER_MESSAGE,
      '',
      ...buildErrorPosition(getInputAsString(this.mostRecentInput), this.pos),
    ].join('\n');
  }
}

File: proposed-projects/round-trip-experiment/reverse/20260909T180307Z_280f3890/20260910T095157Z_2fde87b4/ported_implementation/gbnf/utils/errors/input_parse_error.py
from .build_error_position import build_error_position
from .errors_types import ValidInput
from .get_input_as_string import get_input_as_string

INPUT_PARSER_ERROR_HEADER_MESSAGE = "Failed to parse input string:"


def _build_message(src: str, pos: int) -> str:
    return "\n".join(
        [INPUT_PARSER_ERROR_HEADER_MESSAGE, "", *build_error_position(src, pos)]
    )


class InputParseError(Exception):
    def __init__(
        self, most_recent_input: ValidInput, pos: int, previous_input: ValidInput = ""
    ):
        previous = get_input_as_string(previous_input)
        message = _build_message(
            f"{previous}{get_input_as_string(most_recent_input)}", pos + len(previous)
        )
        super().__init__(message)
        self.name = "InputParseError"
        self.message = message
        self.most_recent_input = most_recent_input
        self.pos = pos
        self.previous_input = previous_input

    @property
    def src(self) -> str:
        return (
            f"{get_input_as_string(self.previous_input)}"
            f"{get_input_as_string(self.most_recent_input)}"
        )

    @property
    def error_for_most_recent_input(self) -> str:
        return _build_message(get_input_as_string(self.most_recent_input), self.pos)

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, InputParseError):
            return NotImplemented
        return str(self) == str(other)

    __hash__ = object.__hash__
